File: ECE39500/exams/midterm-2/PartA/AddState.java
public class AddState extends CompilerState {

   Compiler compiler = null; 

   public AddState(Compiler compiler) {
      this.compiler = compiler;
   }

   public void comma( ) { 
      System.out.println("Parsing Add, expected int, found ,");
      compiler.state = StateFactory.getState("ErrorState");
   }

   public void value(String str) {
      compiler.values[++compiler.valueCount] = Integer.parseInt(str);
      compiler.state = StateFactory.getState("CVCVState");
   }

   public void eol( ) { 
      if (validOperands(compiler.values, 3)) {
         System.out.print("reg["+compiler.values[2]+"] = reg[");
         System.out.print(compiler.values[0]+"] + reg[");
         System.out.println(compiler.values[1]+"]\n");
      } else {
         printInvalidOperands(compiler.values, 3);
      }
      compiler.state = StateFactory.getState("InitState");
   }
      
   public void eof( ) {
      System.out.println("Error in add, expected value, found eof");
      compiler.state = StateFactory.getState("ErrorState");
   }

   public String toString( ) {
      return "AddState";
   }
}

File: ECE39500/exams/midterm-2/midterm-2/PartA/AddState.py
from CompilerState import CompilerState
from StateFactory import StateFactory
import sys

class AddState(CompilerState):
    INT_MIN = -sys.maxint - 1

    def __init__(self, compiler):
        self.compiler = compiler
        return
    
    def comma(self):
        self.compiler.state = StateFactory.getState("ErrorState")

    def value(self, s):
        self.compiler.valueCount += 1
        self.compiler.values[self.compiler.valueCount] = int(str)
        self.compiler.state = StateFactory.getState("CVCVState")

    def eol(self):
        if (self.validOperands(self.compiler.values, 3)):
            print(f"reg[{self.compiler.values[2]}] = reg[{self.compiler.values[0]}] + reg[{self.compiler.values[1]}]")
        else:
            self.printInvalidOperands(self.compiler.values, 3)
        self.compiler.state = StateFactory.getState("InitState")

    def eof(self):
        print("Parsing AddState, expected value, found eof")
        self.compiler.state = StateFactory.getState("ErrorState")
    
    def __str__(self):
        return "AddState"